daemon/src/lib.rs

```diff
@@ -0,0 +1,655 @@
+            Err(error) => Err(anyhow!("{}: open failed ({error})", pidfile.display()))?,
+            Ok(file) => file,
+        };
+
+        match sys::fcntl_get_lock(&file)? {
+            None => {
+                // `pidfile` exists but it isn't locked by anyone.  We need to delete it so that a
+                // new pidfile can go in its place.  But just unlinking it directly makes a nasty
+                // race: what if someone else unlinks it before we do and then replaces it by a
+                // valid pidfile?  We'd unlink their valid pidfile.  We do a little dance to avoid
+                // the race, by locking the invalid pidfile.  Only one process can have the invalid
+                // pidfile locked, and only that process has the right to unlink it.
+                if !delete_if_stale {
+                    Err(anyhow!("{}: pidfile is stale", pidfile.display()))?;
+                }
+
+                // Get the lock.
+                sys::fcntl_set_lock(&file)
+                    .with_context(|| format!("{}: lost race to lock pidfile", pidfile.display()))?;
+
+                // Is the file we have locked still named `pidfile`?
+                match (std::fs::metadata(pidfile), file.metadata()) {
+                    (Ok(m1), Ok(m2)) if m1.dev() == m2.dev() && m1.ino() == m2.ino() => (),
+                    _ => {
+                        // No.  We lost a race with someone else who got the lock before us,
+                        // deleted the pidfile, and closed it (releasing the lock).
+                        Err(anyhow!("{}: lost race to delete pidfile", pidfile.display()))?;
+                    }
+                };
+
+                // We won the right to delete the stale pidfile.
+                std::fs::remove_file(pidfile)
+                    .with_context(|| format!("{}: failed to delete stale pidfile", pidfile.display()))?;
+                Ok(None)
+            },
+            Some(lock_pid) => {
+                let mut reader = BufReader::new(file);
+                let mut line = String::new();
+                match reader.read_line(&mut line) {
+                    Err(e) => Err(e).with_context(|| format!("{}: read failed", pidfile.display()))?,
+                    Ok(0) => Err(anyhow!("{}: read: unexpected end of file", pidfile.display()))?,
+                    Ok(_) => (),
+                };
+                let read_pid: i32 = line.trim().parse()?;
+                if lock_pid != read_pid {
+                    Err(anyhow!("{}: stale pidfile for pid {read_pid} being deleted by id {lock_pid}",
+                                pidfile.display()))?;
+                }
+                Ok(Some(lock_pid))
+            }
+        }
+    }
+}
+
+
+fn count_threads(pid: u32) -> Result<usize, Error> {
+    Ok(read_dir(format!("/proc/{pid}/task"))?.count())
+}
+
+fn assert_single_threaded() {
+    assert_eq!(count_threads(std::process::id()).unwrap(), 1);
```

Comment:
this is also racy, but I don't think you can do any better.

Reply:
It is not racy for this purpose, because if there is just one thread, then we are that thread, and we are not starting a new thread at the moment.